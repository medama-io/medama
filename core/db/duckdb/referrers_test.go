package duckdb_test

import (
	"testing"

	"github.com/gkampitakis/go-snaps/snaps"
)

func TestGetWebsiteReferrersSummary(t *testing.T) {
	_, require, ctx, client := UseDatabaseFixture(t, SIMPLE_FIXTURE)

	testCases := getBaseTestCases(MEDIUM_HOSTNAME)

	for _, tc := range testCases {
		t.Run(tc.Name, func(t *testing.T) {
			summary, err := client.GetWebsiteReferrersSummary(ctx, tc.Filters)
			require.NoError(err)

			snap := NewSnapRecords(summary)
			snaps.MatchSnapshot(t, snap.Snapshot())
		})
	}
}

func TestGetWebsiteReferrers(t *testing.T) {
	_, require, ctx, client := UseDatabaseFixture(t, SIMPLE_FIXTURE)

	testCases := getBaseTestCases(MEDIUM_HOSTNAME)

	for _, tc := range testCases {
		t.Run(tc.Name, func(t *testing.T) {
			referrers, err := client.GetWebsiteReferrers(ctx, tc.Filters)
			require.NoError(err)

			snap := NewSnapRecords(referrers)
			snaps.MatchSnapshot(t, snap.Snapshot())
		})
	}
}
