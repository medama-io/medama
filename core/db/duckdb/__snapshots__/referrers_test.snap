
[TestGetWebsiteReferrers/Base - 1]
RECORDS:
&{StatsReferrerSummary:{Referrer:google.com Visitors:167163 VisitorsPercentage:0.3346} BounceRate:0.487 Duration:5018}
&{StatsReferrerSummary:{Referrer:medama.io Visitors:166318 VisitorsPercentage:0.3329} BounceRate:0.4931 Duration:4985}
&{StatsReferrerSummary:{Referrer: Visitors:166176 VisitorsPercentage:0.3326} BounceRate:0.4885 Duration:5005}

---

[TestGetWebsiteReferrers/Browser - 1]
RECORDS:
&{StatsReferrerSummary:{Referrer:medama.io Visitors:55495 VisitorsPercentage:0.3341} BounceRate:0.4955 Duration:4944}
&{StatsReferrerSummary:{Referrer: Visitors:55307 VisitorsPercentage:0.333} BounceRate:0.4888 Duration:4980}
&{StatsReferrerSummary:{Referrer:google.com Visitors:55288 VisitorsPercentage:0.3329} BounceRate:0.4888 Duration:5016}

---

[TestGetWebsiteReferrers/Country - 1]
RECORDS:
&{StatsReferrerSummary:{Referrer: Visitors:18505 VisitorsPercentage:0.335} BounceRate:0.4867 Duration:5008}
&{StatsReferrerSummary:{Referrer:google.com Visitors:18444 VisitorsPercentage:0.3339} BounceRate:0.4881 Duration:5025}
&{StatsReferrerSummary:{Referrer:medama.io Visitors:18294 VisitorsPercentage:0.3312} BounceRate:0.4912 Duration:4938}

---

[TestGetWebsiteReferrers/Device - 1]
RECORDS:
&{StatsReferrerSummary:{Referrer: Visitors:6264 VisitorsPercentage:0.3357} BounceRate:0.476 Duration:4975}
&{StatsReferrerSummary:{Referrer:google.com Visitors:6262 VisitorsPercentage:0.3356} BounceRate:0.4772 Duration:5099}
&{StatsReferrerSummary:{Referrer:medama.io Visitors:6135 VisitorsPercentage:0.3288} BounceRate:0.5029 Duration:4901}

---

[TestGetWebsiteReferrers/Empty - 1]
RECORDS:

---

[TestGetWebsiteReferrers/Language - 1]
RECORDS:
&{StatsReferrerSummary:{Referrer: Visitors:3189 VisitorsPercentage:0.3362} BounceRate:0.4789 Duration:4998}
&{StatsReferrerSummary:{Referrer:google.com Visitors:3159 VisitorsPercentage:0.333} BounceRate:0.4686 Duration:5096}
&{StatsReferrerSummary:{Referrer:medama.io Visitors:3138 VisitorsPercentage:0.3308} BounceRate:0.4997 Duration:4909}

---

[TestGetWebsiteReferrers/OS - 1]
RECORDS:
&{StatsReferrerSummary:{Referrer: Visitors:1055 VisitorsPercentage:0.3386} BounceRate:0.5159 Duration:4735}
&{StatsReferrerSummary:{Referrer:medama.io Visitors:1036 VisitorsPercentage:0.3325} BounceRate:0.5087 Duration:4809}
&{StatsReferrerSummary:{Referrer:google.com Visitors:1025 VisitorsPercentage:0.3289} BounceRate:0.4477 Duration:5257}

---

[TestGetWebsiteReferrers/Pathname - 1]
RECORDS:
&{StatsReferrerSummary:{Referrer:medama.io Visitors:381 VisitorsPercentage:0.3495} BounceRate:0.5196 Duration:4728}
&{StatsReferrerSummary:{Referrer:google.com Visitors:355 VisitorsPercentage:0.3257} BounceRate:0.45 Duration:5308}
&{StatsReferrerSummary:{Referrer: Visitors:354 VisitorsPercentage:0.3248} BounceRate:0.4908 Duration:5411}

---

[TestGetWebsiteReferrers/Referrer - 1]
RECORDS:
&{StatsReferrerSummary:{Referrer:medama.io Visitors:381 VisitorsPercentage:1} BounceRate:0.5196 Duration:4728}

---

[TestGetWebsiteReferrers/UTMCampaign - 1]
RECORDS:
&{StatsReferrerSummary:{Referrer:medama.io Visitors:116 VisitorsPercentage:1} BounceRate:0.5439 Duration:4732}

---

[TestGetWebsiteReferrers/UTMMedium - 1]
RECORDS:
&{StatsReferrerSummary:{Referrer:medama.io Visitors:35 VisitorsPercentage:1} BounceRate:0.4615 Duration:6547}

---

[TestGetWebsiteReferrers/UTMSource - 1]
RECORDS:
&{StatsReferrerSummary:{Referrer:medama.io Visitors:17 VisitorsPercentage:1} BounceRate:0 Duration:3389}

---

[TestGetWebsiteReferrersGrouped/Base - 1]
RECORDS:
&{StatsReferrerSummary:{Referrer:google.com Visitors:167163 VisitorsPercentage:0.3346} BounceRate:0.487 Duration:5018}
&{StatsReferrerSummary:{Referrer:medama.io Visitors:166318 VisitorsPercentage:0.3329} BounceRate:0.4931 Duration:4985}
&{StatsReferrerSummary:{Referrer: Visitors:166176 VisitorsPercentage:0.3326} BounceRate:0.4885 Duration:5005}

---

[TestGetWebsiteReferrersGrouped/Browser - 1]
RECORDS:
&{StatsReferrerSummary:{Referrer:medama.io Visitors:55495 VisitorsPercentage:0.3341} BounceRate:0.4955 Duration:4944}
&{StatsReferrerSummary:{Referrer: Visitors:55307 VisitorsPercentage:0.333} BounceRate:0.4888 Duration:4980}
&{StatsReferrerSummary:{Referrer:google.com Visitors:55288 VisitorsPercentage:0.3329} BounceRate:0.4888 Duration:5016}

---

[TestGetWebsiteReferrersGrouped/Country - 1]
RECORDS:
&{StatsReferrerSummary:{Referrer: Visitors:18505 VisitorsPercentage:0.335} BounceRate:0.4867 Duration:5008}
&{StatsReferrerSummary:{Referrer:google.com Visitors:18444 VisitorsPercentage:0.3339} BounceRate:0.4881 Duration:5025}
&{StatsReferrerSummary:{Referrer:medama.io Visitors:18294 VisitorsPercentage:0.3312} BounceRate:0.4912 Duration:4938}

---

[TestGetWebsiteReferrersGrouped/Device - 1]
RECORDS:
&{StatsReferrerSummary:{Referrer: Visitors:6264 VisitorsPercentage:0.3357} BounceRate:0.476 Duration:4975}
&{StatsReferrerSummary:{Referrer:google.com Visitors:6262 VisitorsPercentage:0.3356} BounceRate:0.4772 Duration:5099}
&{StatsReferrerSummary:{Referrer:medama.io Visitors:6135 VisitorsPercentage:0.3288} BounceRate:0.5029 Duration:4901}

---

[TestGetWebsiteReferrersGrouped/Empty - 1]
RECORDS:

---

[TestGetWebsiteReferrersGrouped/Language - 1]
RECORDS:
&{StatsReferrerSummary:{Referrer: Visitors:3189 VisitorsPercentage:0.3362} BounceRate:0.4789 Duration:4998}
&{StatsReferrerSummary:{Referrer:google.com Visitors:3159 VisitorsPercentage:0.333} BounceRate:0.4686 Duration:5096}
&{StatsReferrerSummary:{Referrer:medama.io Visitors:3138 VisitorsPercentage:0.3308} BounceRate:0.4997 Duration:4909}

---

[TestGetWebsiteReferrersGrouped/OS - 1]
RECORDS:
&{StatsReferrerSummary:{Referrer: Visitors:1055 VisitorsPercentage:0.3386} BounceRate:0.5159 Duration:4735}
&{StatsReferrerSummary:{Referrer:medama.io Visitors:1036 VisitorsPercentage:0.3325} BounceRate:0.5087 Duration:4809}
&{StatsReferrerSummary:{Referrer:google.com Visitors:1025 VisitorsPercentage:0.3289} BounceRate:0.4477 Duration:5257}

---

[TestGetWebsiteReferrersGrouped/Pathname - 1]
RECORDS:
&{StatsReferrerSummary:{Referrer:medama.io Visitors:381 VisitorsPercentage:0.3495} BounceRate:0.5196 Duration:4728}
&{StatsReferrerSummary:{Referrer:google.com Visitors:355 VisitorsPercentage:0.3257} BounceRate:0.45 Duration:5308}
&{StatsReferrerSummary:{Referrer: Visitors:354 VisitorsPercentage:0.3248} BounceRate:0.4908 Duration:5411}

---

[TestGetWebsiteReferrersGrouped/Referrer - 1]
RECORDS:
&{StatsReferrerSummary:{Referrer:medama.io Visitors:381 VisitorsPercentage:1} BounceRate:0.5196 Duration:4728}

---

[TestGetWebsiteReferrersGrouped/UTMCampaign - 1]
RECORDS:
&{StatsReferrerSummary:{Referrer:medama.io Visitors:116 VisitorsPercentage:1} BounceRate:0.5439 Duration:4732}

---

[TestGetWebsiteReferrersGrouped/UTMMedium - 1]
RECORDS:
&{StatsReferrerSummary:{Referrer:medama.io Visitors:35 VisitorsPercentage:1} BounceRate:0.4615 Duration:6547}

---

[TestGetWebsiteReferrersGrouped/UTMSource - 1]
RECORDS:
&{StatsReferrerSummary:{Referrer:medama.io Visitors:17 VisitorsPercentage:1} BounceRate:0 Duration:3389}

---

[TestGetWebsiteReferrersSummary/Base - 1]
RECORDS:
&{Referrer:google.com Visitors:167163 VisitorsPercentage:0.3346}
&{Referrer:medama.io Visitors:166318 VisitorsPercentage:0.3329}
&{Referrer: Visitors:166176 VisitorsPercentage:0.3326}

---

[TestGetWebsiteReferrersSummary/Browser - 1]
RECORDS:
&{Referrer:medama.io Visitors:55495 VisitorsPercentage:0.3341}
&{Referrer: Visitors:55307 VisitorsPercentage:0.333}
&{Referrer:google.com Visitors:55288 VisitorsPercentage:0.3329}

---

[TestGetWebsiteReferrersSummary/Country - 1]
RECORDS:
&{Referrer: Visitors:18505 VisitorsPercentage:0.335}
&{Referrer:google.com Visitors:18444 VisitorsPercentage:0.3339}
&{Referrer:medama.io Visitors:18294 VisitorsPercentage:0.3312}

---

[TestGetWebsiteReferrersSummary/Device - 1]
RECORDS:
&{Referrer: Visitors:6264 VisitorsPercentage:0.3357}
&{Referrer:google.com Visitors:6262 VisitorsPercentage:0.3356}
&{Referrer:medama.io Visitors:6135 VisitorsPercentage:0.3288}

---

[TestGetWebsiteReferrersSummary/Empty - 1]
RECORDS:

---

[TestGetWebsiteReferrersSummary/Language - 1]
RECORDS:
&{Referrer: Visitors:3189 VisitorsPercentage:0.3362}
&{Referrer:google.com Visitors:3159 VisitorsPercentage:0.333}
&{Referrer:medama.io Visitors:3138 VisitorsPercentage:0.3308}

---

[TestGetWebsiteReferrersSummary/OS - 1]
RECORDS:
&{Referrer: Visitors:1055 VisitorsPercentage:0.3386}
&{Referrer:medama.io Visitors:1036 VisitorsPercentage:0.3325}
&{Referrer:google.com Visitors:1025 VisitorsPercentage:0.3289}

---

[TestGetWebsiteReferrersSummary/Pathname - 1]
RECORDS:
&{Referrer:medama.io Visitors:381 VisitorsPercentage:0.3495}
&{Referrer:google.com Visitors:355 VisitorsPercentage:0.3257}
&{Referrer: Visitors:354 VisitorsPercentage:0.3248}

---

[TestGetWebsiteReferrersSummary/Referrer - 1]
RECORDS:
&{Referrer:medama.io Visitors:381 VisitorsPercentage:1}

---

[TestGetWebsiteReferrersSummary/UTMCampaign - 1]
RECORDS:
&{Referrer:medama.io Visitors:116 VisitorsPercentage:1}

---

[TestGetWebsiteReferrersSummary/UTMMedium - 1]
RECORDS:
&{Referrer:medama.io Visitors:35 VisitorsPercentage:1}

---

[TestGetWebsiteReferrersSummary/UTMSource - 1]
RECORDS:
&{Referrer:medama.io Visitors:17 VisitorsPercentage:1}

---

[TestGetWebsiteReferrersSummaryGrouped/Base - 1]
RECORDS:
&{Referrer:google.com Visitors:167163 VisitorsPercentage:0.3346}
&{Referrer:medama.io Visitors:166318 VisitorsPercentage:0.3329}
&{Referrer: Visitors:166176 VisitorsPercentage:0.3326}

---

[TestGetWebsiteReferrersSummaryGrouped/Browser - 1]
RECORDS:
&{Referrer:medama.io Visitors:55495 VisitorsPercentage:0.3341}
&{Referrer: Visitors:55307 VisitorsPercentage:0.333}
&{Referrer:google.com Visitors:55288 VisitorsPercentage:0.3329}

---

[TestGetWebsiteReferrersSummaryGrouped/Country - 1]
RECORDS:
&{Referrer: Visitors:18505 VisitorsPercentage:0.335}
&{Referrer:google.com Visitors:18444 VisitorsPercentage:0.3339}
&{Referrer:medama.io Visitors:18294 VisitorsPercentage:0.3312}

---

[TestGetWebsiteReferrersSummaryGrouped/Device - 1]
RECORDS:
&{Referrer: Visitors:6264 VisitorsPercentage:0.3357}
&{Referrer:google.com Visitors:6262 VisitorsPercentage:0.3356}
&{Referrer:medama.io Visitors:6135 VisitorsPercentage:0.3288}

---

[TestGetWebsiteReferrersSummaryGrouped/Empty - 1]
RECORDS:

---

[TestGetWebsiteReferrersSummaryGrouped/Language - 1]
RECORDS:
&{Referrer: Visitors:3189 VisitorsPercentage:0.3362}
&{Referrer:google.com Visitors:3159 VisitorsPercentage:0.333}
&{Referrer:medama.io Visitors:3138 VisitorsPercentage:0.3308}

---

[TestGetWebsiteReferrersSummaryGrouped/OS - 1]
RECORDS:
&{Referrer: Visitors:1055 VisitorsPercentage:0.3386}
&{Referrer:medama.io Visitors:1036 VisitorsPercentage:0.3325}
&{Referrer:google.com Visitors:1025 VisitorsPercentage:0.3289}

---

[TestGetWebsiteReferrersSummaryGrouped/Pathname - 1]
RECORDS:
&{Referrer:medama.io Visitors:381 VisitorsPercentage:0.3495}
&{Referrer:google.com Visitors:355 VisitorsPercentage:0.3257}
&{Referrer: Visitors:354 VisitorsPercentage:0.3248}

---

[TestGetWebsiteReferrersSummaryGrouped/Referrer - 1]
RECORDS:
&{Referrer:medama.io Visitors:381 VisitorsPercentage:1}

---

[TestGetWebsiteReferrersSummaryGrouped/UTMCampaign - 1]
RECORDS:
&{Referrer:medama.io Visitors:116 VisitorsPercentage:1}

---

[TestGetWebsiteReferrersSummaryGrouped/UTMMedium - 1]
RECORDS:
&{Referrer:medama.io Visitors:35 VisitorsPercentage:1}

---

[TestGetWebsiteReferrersSummaryGrouped/UTMSource - 1]
RECORDS:
&{Referrer:medama.io Visitors:17 VisitorsPercentage:1}

---
